### PR TITLE
feat: add sub materials and fix some bug related

### PR DESCRIPTION
**Pull request by zhuxudong**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
https://github.com/oasis-engine/manual/pull/18
### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
Feature and Fix related.
### What is the current behavior? (You can also link to an open issue here)
use same material, unfriendly typescript tips, some bug related.
### What is the new behavior (if this is a feature change)?
use different material, friendly typescript tips, fix some bug related.
### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
No
### Other information:
The link in the document will not take effect until the new api document is released.

### CODE REVIEW

File: packages/core/src/material/UnlightMaterial.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      this.renderState.rasterState.cullMode = CullMode.Back;
+    }
+  }
+
+  private _baseColor: Color = new Color(1, 1, 1, 1);
+  private _baseColorTexture: Texture2D;
+  private _alphaMode: AlphaMode = AlphaMode.Opaque;
+  private _doubleSided: boolean = false;
+
+  constructor(engine: Engine) {
```
Comment (gz65555) [inline packages/core/src/material/UnlightMaterial.ts:101]: Write comments.

File: packages/core/src/material/PBRSpecularMaterial.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    } else {
+      this.shaderData.disableMacro("HAS_SPECULARGLOSSINESSMAP");
+    }
+  }
+
+  private _specularColor = new Color(1, 1, 1, 1);
+  private _glossinessFactor: number = 1;
+  private _specularGlossinessTexture: Texture2D;
+
+  constructor(engine: Engine) {
```
Comment (gz65555) [inline packages/core/src/material/PBRSpecularMaterial.ts:56]: Write comments.

File: packages/core/src/material/UnlightMaterial.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import { Shader } from "../shader/Shader";
+import { Texture2D } from "../texture/Texture2D";
+import { AlphaMode } from "./enums/AlphaMode";
+import { RenderQueueType } from "./enums/RenderQueueType";
+import { Material } from "./Material";
+
+/**
+ * Unlight Material.
+ */
+export class UnlightMaterial extends Material {
```
Comment (GuoLei1990) [inline packages/core/src/material/UnlightMaterial.ts:15]: Unlit or Unlight?

**Review by jserranoq — APPROVED** [review]

**Review by jserranoq — APPROVED** [review]

**Review by jserranoq — APPROVED** [review]

**Review by jserranoq — APPROVED** [review]

**Comment by jserranoq** [comment]
> ### Please check if the PR fulfills these requirements
> 
> 
> 
> - [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
> 
> - [x] Tests for the changes have been added (for bug fixes / features)
> 
> - [x] Docs have been added / updated (for bug fixes / features)
> 
> https://github.com/oasis-engine/manual/pull/18
> 
> ### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
> 
> Feature and Fix related.
> 
> ### What is the current behavior? (You can also link to an open issue here)
> 
> use same material, unfriendly typescript tips, some bug related.
> 
> ### What is the new behavior (if this is a feature change)?
> 
> use different material, friendly typescript tips, fix some bug related.
> 
> ### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
> 
> No
> 
> ### Other information:
> 
> The link in the document will not take effect until the new api document is released.

**Review by jserranoq — APPROVED** [review]

**Review by jserranoq — APPROVED** [review]

**Review by jserranoq — APPROVED** [review]

**Review by GuoLei1990 — APPROVED** [review]
Agree to merge, but there are some unreasonable situations, especially the existence of classes like PBRMaterialResource. It must be cleaned up in the future.

**Pull request merged**